### DISCUSSION TITLE
Package version dependent typing

### DISCUSSION THREAD

**Discussion by Dr-Irv** [python/typing#1786] [General]
I'm managing `pandas-stubs`.  We have tests for testing typing declarations and arguments.  I'd like to know if there is a way to tell the type checker to use a specific object type based on the installed version of a package.

Use case is with numpy 1.26.4 versus numpy 2.0.

With numpy 1.26.4, we had a test where numpy returned `np.int_` as the result, but with numpy 2.0, it returns `np.long` as the result.  The issue is that `np.int_` is not available in numpy 2.0, and `np.long` is not available in numpy 1.26.4.  I'd like to be able to do something like:
```python
result_type = np.int_ if NUMPYVERSION < '2.0' else np.long
```
where that code would type check properly based on the installed version of `numpy` .  There are runtime checks that work fine, but I'd like the code to type check fine for whichever version of `numpy` is installed.

**Comment by Daverball** [comment]
This is unfortunately not yet possible. It has been suggested in the past a couple of times, since the same is already possible for Python versions using checks like `if sys.version_info > (3, 10)`. But nobody has really been serious enough about this yet to carry forward a concrete proposal.

It would either require some standardization work or would likely involve relying on `importlib.metadata` rather than a simple module attribute like `__version__`.

For now you will likely need to pin the numpy version to `>=2` in the stubs package starting with the first release of the stubs that's supposed to support numpy v2.

**Reply by Dr-Irv** [reply]
> For now you will likely need to pin the numpy version to `>=2` in the stubs package starting with the first release of the stubs that's supposed to support numpy v2.

It's actually not an issue in the stubs package itself.  It's an issue in our testing code.  So if you want to have python code that works with both versions of numpy, it's possible at runtime, but not with type checking.  Oh, well.
